File: frontend/types/dataset.ts
export type DatasetStatus =
  "uploading" | "validating" | "ready" | "failed" | "archived";

export interface DatasetWarning {
  code: string;
  message: string;
}
export interface DatasetSummary {
  id: string;
  original_filename: string;
  file_extension: string;
  mime_type: string;
  file_size_bytes: number;
  row_count: number;
  column_count: number;
  status: DatasetStatus;
  created_at: string;
  preview_available: boolean;
}
export interface DatasetDetail extends DatasetSummary {
  checksum_sha256: string;
  column_names: string[];
  delimiter: string | null;
  encoding: string | null;
  updated_at: string;
  deleted_at: string | null;
  ingestion_version: number;
  source_type: string;
  warnings: DatasetWarning[];
}
export interface DatasetUploadResponse extends DatasetDetail {
  preview_rows: Array<Record<string, string | null>>;
}
export interface DatasetPreview {
  dataset_id: string;
  columns: string[];
  rows: Array<Record<string, string | null>>;
  returned_rows: number;
  max_rows: number;
}
export interface PaginationMetadata {
  page: number;
  page_size: number;
  total_items: number;
  total_pages: number;
}
export interface DatasetListResponse {
  items: DatasetSummary[];
  pagination: PaginationMetadata;
}
export interface DatasetArchiveResponse {
  id: string;
  status: "archived";
  deleted_at: string;
}
export interface DatasetStats {
  active_datasets: number;
  latest_filename: string | null;
  latest_upload_at: string | null;
  ingestion_status: "operational";
}
export interface DatasetApiErrorBody {
  detail: string;
  existing_dataset_id?: string;
}

File: backend/app/schemas/dataset.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

DatasetStatusValue = Literal["uploading", "validating", "ready", "failed", "archived"]


class DatasetWarning(BaseModel):
    code: str
    message: str


class DatasetSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    original_filename: str
    file_extension: str
    mime_type: str
    file_size_bytes: int
    row_count: int
    column_count: int
    status: DatasetStatusValue
    created_at: datetime
    preview_available: bool


class DatasetDetail(DatasetSummary):
    checksum_sha256: str
    column_names: list[str]
    delimiter: str | None
    encoding: str | None
    updated_at: datetime
    deleted_at: datetime | None
    ingestion_version: int
    source_type: str
    warnings: list[DatasetWarning] = []


class DatasetUploadResponse(DatasetDetail):
    preview_rows: list[dict[str, str | None]]


class DatasetPreviewResponse(BaseModel):
    dataset_id: str
    columns: list[str]
    rows: list[dict[str, str | None]]
    returned_rows: int
    max_rows: int


class PaginationMetadata(BaseModel):
    page: int
    page_size: int
    total_items: int
    total_pages: int


class DatasetListResponse(BaseModel):
    items: list[DatasetSummary]
    pagination: PaginationMetadata


class DatasetArchiveResponse(BaseModel):
    id: str
    status: Literal["archived"]
    deleted_at: datetime


class DuplicateDatasetResponse(BaseModel):
    detail: str
    existing_dataset_id: str


class DatasetErrorResponse(BaseModel):
    detail: str


class DatasetStatsResponse(BaseModel):
    active_datasets: int = Field(ge=0)
    latest_filename: str | None
    latest_upload_at: datetime | None
    ingestion_status: Literal["operational"] = "operational"
